source/common/router/filter_config_discovery_impl.h

```diff
@@ -0,0 +1,170 @@
+
+  // Router::FilterConfigProvider
+  const std::string& name() override;
+  absl::optional<Http::FilterFactoryCb> config() override;
+  void validateConfig(Server::Configuration::NamedHttpFilterConfigFactory& factory) override;
+  void onConfigUpdate(Http::FilterFactoryCb config, const std::string&) override;
+
+private:
+  struct ThreadLocalConfig : public ThreadLocal::ThreadLocalObject {
+    ThreadLocalConfig() : config_{absl::nullopt} {}
+    absl::optional<Http::FilterFactoryCb> config_{};
+  };
+
+  FilterConfigSubscriptionSharedPtr subscription_;
+  const bool require_terminal_;
+  ThreadLocal::SlotPtr tls_;
+
+  // Local initialization target to ensure that the subscription starts in
+  // case no warming is requested by any other filter config provider.
+  Init::TargetImpl init_target_;
+
+  friend class FilterConfigProviderManagerImpl;
+};
+
+/**
+ * A class that fetches the filter configuration dynamically using the filter config discovery API.
+ * Subscriptions are shared between the filter config providers. The filter config providers are
+ * notified when a new config is accepted.
+ */
+class FilterConfigSubscription
+    : Envoy::Config::SubscriptionBase<envoy::config::core::v3::TypedExtensionConfig>,
+      Logger::Loggable<Logger::Id::router> {
+public:
+  FilterConfigSubscription(const envoy::config::core::v3::ConfigSource& config_source,
+                           const std::string& filter_config_name,
+                           Server::Configuration::FactoryContext& factory_context,
+                           const std::string& stat_prefix,
+                           FilterConfigProviderManagerImpl& filter_config_provider_manager,
+                           const std::string& subscription_id);
+
+  ~FilterConfigSubscription() override;
+
+  const Init::SharedTargetImpl& initTarget() { return parent_init_target_; }
+  const std::string& name() { return filter_config_name_; }
+
+private:
+  void start();
+
+  // Config::SubscriptionCallbacks
+  void onConfigUpdate(const std::vector<Envoy::Config::DecodedResourceRef>& resources,
+                      const std::string& version_info) override;
+  void onConfigUpdate(const std::vector<Envoy::Config::DecodedResourceRef>& added_resources,
+                      const Protobuf::RepeatedPtrField<std::string>& removed_resources,
+                      const std::string&) override;
+  void onConfigUpdateFailed(Envoy::Config::ConfigUpdateFailureReason reason,
+                            const EnvoyException*) override;
+
+  std::unique_ptr<Envoy::Config::Subscription> subscription_;
+  const std::string filter_config_name_;
+  Server::Configuration::FactoryContext& factory_context_;
+  ProtobufMessage::ValidationVisitor& validator_;
+
+  // Parent init target initializes local init manager, which in turn
+  // initializes local init target to start a subscription. The subscription
+  // signals the watcher to notify the parent init target readiness.
```

Comment:
This is sort of what I was asking for, but it's still not very clear to me. Why is there a local init manager, i.e. what is the transitive problem being solved.

Reply:
Yeah, there's no transitive requirement, so it's gone.